### PR TITLE
kmeans update

### PR DESCRIPTION
**Pull request by Saurabh7**
This has most of the primary changes to Kmeans based on https://github.com/shogun-toolbox/shogun/issues/2987 and some general cleanups.
Still have to do some aesthetic changes.

- Changed random initialization
  - now chooses `k` random points as centers rather than assigning random centers to all points.
  - Removed unnecessary paramters, cluster_list, weight_set, etc. This should be calculated in the Lloyds implementation.
- changed manual initialization
  - again Removed unnecessary paramters
  - removed calculation of  distances and assigning of clusters from here, this along wth the parameters now takes place in assignment step of lloyds.

- made clear separation for assigment step and update step in lloyds impl.
- removed random sampling of instances to be updated in each iteration, each sample should be updated. 
- used precomputed squared norms for euclidean distances
- removed various unnecessary computations in lloyds impl
  - no need to change center values in assigment steps now takes place only in update step.
  - no need to store distances in iterations, now calculated in place (scope for parallelisation) 
- some new unit-tests

### CODE REVIEW

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 1]
```diff
@@ -37,107 +49,117 @@ void CKMeansLloydImpl::Lloyd_KMeans(int32_t k, CDistance* distance, int32_t max_
...
-				/* weights_set(imini) = weights_set(imini) + 1.0 ; */
-				weights_set[imini]+= 1.0;
-				/* weights_set(j)     = weights_set(j)     - 1.0 ; */
-				weights_set[ClList_Pat]-= 1.0;
-
-				vec=lhs->get_feature_vector(Pat, vlen, vfree);
-
-				for (j=0; j<dimensions; j++)
-				{
-					mus.matrix[imini*dimensions+j]-=
```
Comment (Saurabh7) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: I believe this was actually completely wrong, should be +=

File: src/shogun/clustering/KMeans.cpp [hunk 2]
```diff
@@ -67,113 +66,41 @@ CKMeans::~CKMeans()
...
-	for (int32_t i=0; i<XSize; i++)
-	{
-		const int32_t Cl = ClList[i];
-		weights_set[Cl]+=1.0;
-
+		const int32_t Cl=temp[i];
 		int32_t vlen=0;
 		bool vfree=false;
-		float64_t* vec=lhs->get_feature_vector(i, vlen, vfree);
+		float64_t* vec=lhs->get_feature_vector(Cl, vlen, vfree);
```
Comment (karlnapf) [inline src/shogun/clustering/KMeans.cpp]: can you use the SGVector version of get_feature vector here?

File: src/shogun/clustering/KMeans.cpp [hunk 3]
```diff
@@ -67,113 +66,41 @@ CKMeans::~CKMeans()
...
-	/* Compute the sum of all points belonging to a cluster
-	 * and count the points */
-	CDenseFeatures<float64_t>* lhs=
-			(CDenseFeatures<float64_t>*)distance->get_lhs();
-	for (int32_t i=0; i<XSize; i++)
-	{
-		const int32_t Cl = ClList[i];
-		weights_set[Cl]+=1.0;
-
+		const int32_t Cl=temp[i];
```
Comment (karlnapf) [inline src/shogun/clustering/KMeans.cpp]: if you start touching things, we dont use Capital variable names...Also please give this a name that documents itself, i.e. "class"

File: src/shogun/clustering/KMeans.cpp [hunk 4]
```diff
@@ -67,113 +66,41 @@ CKMeans::~CKMeans()
...
 		int32_t vlen=0;
 		bool vfree=false;
-		float64_t* vec=lhs->get_feature_vector(i, vlen, vfree);
+		float64_t* vec=lhs->get_feature_vector(Cl, vlen, vfree);
 
 		for (int32_t j=0; j<dimensions; j++)
-			mus.matrix[Cl*dimensions+j] += vec[j];
+			mus.matrix[i*dimensions+j] += vec[j];
```
Comment (karlnapf) [inline src/shogun/clustering/KMeans.cpp]: matrix has () operator, which you should use. Why += ?

File: src/shogun/clustering/KMeans.cpp [hunk 5]
```diff
@@ -236,11 +163,11 @@ bool CKMeans::train_machine(CFeatures* data)
 		CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
 
 	ASSERT(lhs);
-	int32_t XSize=lhs->get_num_vectors();
+	int32_t lhs_size=lhs->get_num_vectors();
 	dimensions=lhs->get_num_features();
 	const int32_t XDimk=dimensions*k;
 
-	ASSERT(XSize>0 && dimensions>0);
+	ASSERT(lhs_size>0 && dimensions>0);
```
Comment (karlnapf) [inline src/shogun/clustering/KMeans.cpp]: can you make this a require, and also provide proper error messages

https://github.com/shogun-toolbox/shogun/wiki/Assertions

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 6]
```diff
@@ -17,19 +18,30 @@ using namespace shogun;
...
+	int32_t lhs_size=lhs->get_num_vectors();
 	int32_t dimensions=lhs->get_num_features();
 
 	CDenseFeatures<float64_t>* rhs_mus=new CDenseFeatures<float64_t>(0);
 	CFeatures* rhs_cache=distance->replace_rhs(rhs_mus);
 
-	SGVector<float64_t> dists=SGVector<float64_t>(k*XSize);
-	dists.zero();
+	/* Cluster assignments for each point */
+	SGVector<int32_t> cl_list=SGVector<int32_t>(lhs_size);
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: please rename to something else, this is not a list and "cl" should be "class"

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 7]
```diff
@@ -17,19 +18,30 @@ using namespace shogun;
...
+	cl_list.zero();
+
+	/* Weights : Number of points in each cluster */
+	SGVector<float64_t> weights_set=SGVector<float64_t>(k);
+	weights_set.zero();
+	/* Initially set all weights for zeroth cluster, Changes in assignement step */
+	weights_set[0]=lhs_size;
+
+	if (distance->get_distance_type() == D_EUCLIDEAN)
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: this is not a good idea.
Sorry I know we had a lengthy discussion, but I did not realise this.
The KMeans class should not use any specific features of the distance class specialization. Rather but a virtual method "precompute" in the base class

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 8]
```diff
@@ -37,107 +49,117 @@ void CKMeansLloydImpl::Lloyd_KMeans(int32_t k, CDistance* distance, int32_t max_
 	bool vfree=false;
 	float64_t* vec=NULL;
 
+	/* Iterations */
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: completely useless comment ;)
please remove

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 9]
```diff
@@ -37,107 +49,117 @@ void CKMeansLloydImpl::Lloyd_KMeans(int32_t k, CDistance* distance, int32_t max_
...
 	while (changed && (iter<max_iter))
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: can you make this a for loop and `break` out of it??

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 10]
```diff
@@ -37,107 +49,117 @@ void CKMeansLloydImpl::Lloyd_KMeans(int32_t k, CDistance* distance, int32_t max_
...
 	{
 		iter++;
 		if (iter==max_iter-1)
 			SG_SWARNING("kmeans clustering changed throughout %d iterations stopping...\n", max_iter-1)
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: Can you make this message understandable?

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 11]
```diff
@@ -37,107 +49,117 @@ void CKMeansLloydImpl::Lloyd_KMeans(int32_t k, CDistance* distance, int32_t max_
...
 
-		if (iter%1000 == 0)
-			SG_SINFO("Iteration[%d/%d]: Assignment of %i patterns changed.\n", iter, max_iter, changed)
 		changed=0;
+		rhs_mus->copy_feature_matrix(mus);
+
+		if (distance->get_distance_type() == D_EUCLIDEAN)
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: see above

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 12]
```diff
@@ -37,107 +49,117 @@ void CKMeansLloydImpl::Lloyd_KMeans(int32_t k, CDistance* distance, int32_t max_
...
 	while (changed && (iter<max_iter))
 	{
 		iter++;
 		if (iter==max_iter-1)
 			SG_SWARNING("kmeans clustering changed throughout %d iterations stopping...\n", max_iter-1)
 
-		if (iter%1000 == 0)
-			SG_SINFO("Iteration[%d/%d]: Assignment of %i patterns changed.\n", iter, max_iter, changed)
 		changed=0;
+		rhs_mus->copy_feature_matrix(mus);
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: this might be memory costly. Is it necessary?

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 13]
```diff
@@ -37,107 +49,117 @@ void CKMeansLloydImpl::Lloyd_KMeans(int32_t k, CDistance* distance, int32_t max_
...
+		{ 
+			const int32_t cl_list_i=cl_list[i];
+			int32_t min_cl, j;
+			float64_t min_dist, dist;
+			
+			min_cl=0;
+		   	min_dist=distance->distance(i,0);
+			for (j=1; j<k; j++)
+			{
+				dist=distance->distance(i,j);
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: we have CMath::max

File: src/shogun/clustering/KMeansLloydImpl.cpp (hunk 1 above)
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: unit tests help for this. Ill write more in a bit

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 14]
```diff
@@ -37,107 +49,117 @@ void CKMeansLloydImpl::Lloyd_KMeans(int32_t k, CDistance* distance, int32_t max_
...
-
-				/* ClList(i)= imini ; */
-				ClList[Pat] = imini;
 			}
+					
 		}
+		
+		if (iter%1000 == 0)
+			SG_SINFO("Iteration[%d/%d]: Assignment of %i patterns changed.\n", iter, max_iter, changed)
+			SG_SPRINT("iter %d, %i", iter, changed)
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: Can you remove the print?

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 15]
```diff
@@ -37,107 +49,117 @@ void CKMeansLloydImpl::Lloyd_KMeans(int32_t k, CDistance* distance, int32_t max_
...
-						mus.matrix[ClList_Pat*dimensions+j]=0;
-				}
-
-				/* ClList(i)= imini ; */
-				ClList[Pat] = imini;
 			}
+					
 		}
+		
+		if (iter%1000 == 0)
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: can you print 10 times during the max iterations?
I.e. if max iterations is 100, you should print every 10
If it is 1000, you should print every 100

File: tests/unit/clustering/kmeans_unittest.cc [hunk 16]
```diff
@@ -162,3 +165,105 @@ TEST(KMeans, minibatch_training_test)
 	SG_UNREF(features);
 }
 
+TEST(KMeans, random_init_fail)
+{
+	/* Create a gaussian with many points. Move very small number of points far away.
+	 * Random initialization should not (very small probability) assign separate
```
Comment (karlnapf) [inline tests/unit/clustering/kmeans_unittest.cc]: please dont do probabilistic unit tests

File: tests/unit/clustering/kmeans_unittest.cc [hunk 17]
```diff
@@ -162,3 +165,105 @@ TEST(KMeans, minibatch_training_test)
...
+			pass++;
+
+		SG_UNREF(learnt_centers);
+	}
+	
+	EXPECT_EQ(pass, 10);
+	SG_UNREF(clustering);
+	SG_UNREF(features);
+	SG_UNREF(gauss)	
+}
```
Comment (karlnapf) [inline tests/unit/clustering/kmeans_unittest.cc]: Unit tests are a quite different idea. These here are too big and complicated, if they fail, they dont tell more of what has failed than an integration test that fails due to a changed result

You should test:
 * corner cases
 * on very small datasets
 * Random initialisation: just test that every center is *exactly* on one data point, and no two centers are the same. How many points and k do you need to test that? Try two cases: k=2, n=3   and k=3, n=3
 * test modules or the code, not chains of multiple modules
 * Dont use random data, unless it leads to *guaranteed* pass

Comment (karlnapf) [inline tests/unit/clustering/kmeans_unittest.cc]: There is another unit test for the random centers, maybe coordinate yourself with that guy, he already put in some work as well

**Comment by karlnapf** [comment]
Thanks for the patch :)
This needs some love until we can merge it, but I look forward to get the fast Kmeans in shogun going

File: src/shogun/clustering/KMeansLloydImpl.cpp (hunk 7 above)
Comment (Saurabh7) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: I had a feeling something was off here, thanks .. added it to base class

File: src/shogun/clustering/KMeansLloydImpl.cpp (hunk 12 above)
Comment (Saurabh7) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: Its necessary, thats how the centers are updated in each iteration.

File: src/shogun/clustering/KMeansLloydImpl.cpp (hunk 13 above)
Comment (Saurabh7) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: yes for that I have to store distances in a vector though. and `min` does similar loops inside.

**Comment by Saurabh7** [comment]
Thanks for the comments ! made changes to make it even better ..

File: tests/unit/clustering/kmeans_unittest.cc [hunk 18]
```diff
@@ -162,3 +165,101 @@ TEST(KMeans, minibatch_training_test)
...
+	CDenseFeatures<float64_t>* features=new CDenseFeatures<float64_t>(rect);
+	SG_REF(features);	
+	CEuclideanDistance* distance=new CEuclideanDistance(features, features);
+	CKMeans* clustering=new CKMeans(2, distance);
+
+	clustering->train(features);
+	CDenseFeatures<float64_t>* learnt_centers=CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
+	SGMatrix<float64_t> c=learnt_centers->get_feature_matrix();
+		
+	EXPECT_EQ(c(0,0), 0);
```
Comment (Saurabh7) [inline tests/unit/clustering/kmeans_unittest.cc]: aah just realized the order can be opposite too , will fix this

Comment (Saurabh7) [inline tests/unit/clustering/kmeans_unittest.cc]: should be fixed now

File: src/shogun/clustering/KMeans.cpp [hunk 19]
```diff
@@ -227,22 +152,24 @@ void CKMeans::compute_cluster_variances()
 
 bool CKMeans::train_machine(CFeatures* data)
 {
-	ASSERT(distance && distance->get_feature_type()==F_DREAL)
-
+	REQUIRE(distance, "Distance is not provided")
+	REQUIRE(distance->get_feature_type()==F_DREAL, "Distance's features should be of type REAL")
```
Comment (karlnapf) [inline src/shogun/clustering/KMeans.cpp]: Distance feature type (%d) should be REAL (%d)

File: src/shogun/clustering/KMeans.cpp [hunk 20]
```diff
@@ -227,22 +152,24 @@ void CKMeans::compute_cluster_variances()
...
+	
 	if (data)
 		distance->init(data, data);
 
 	CDenseFeatures<float64_t>* lhs=
 		CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
 
-	ASSERT(lhs);
-	int32_t XSize=lhs->get_num_vectors();
+	REQUIRE(lhs, "lhs features of distance not provided");
```
Comment (karlnapf) [inline src/shogun/clustering/KMeans.cpp]: Capital L

File: src/shogun/clustering/KMeans.cpp [hunk 21]
```diff
@@ -227,22 +152,24 @@ void CKMeans::compute_cluster_variances()
...
+	int32_t lhs_size=lhs->get_num_vectors();
 	dimensions=lhs->get_num_features();
-	const int32_t XDimk=dimensions*k;
+	const int32_t centers_size=dimensions*k;
 
-	ASSERT(XSize>0 && dimensions>0);
+	REQUIRE(lhs_size>0, "lhs features should not be empty");
```
Comment (karlnapf) [inline src/shogun/clustering/KMeans.cpp]: Capital L.

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 22]
```diff
@@ -17,127 +18,140 @@ using namespace shogun;
...
+	SGVector<int32_t> cluster_assignments=SGVector<int32_t>(lhs_size);
+	cluster_assignments.zero();
+
+	/* Weights : Number of points in each cluster */
+	SGVector<float64_t> weights_set=SGVector<float64_t>(k);
+	weights_set.zero();
+	/* Initially set all weights for zeroth cluster, Changes in assignement step */
+	weights_set[0]=lhs_size;
+
+	distance->precompute_lhs_squared_norms();
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: this method has to get a different name.
I suggest precompute_lhs ... imagine other distances might not be based on lhs norms but something else

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 23]
```diff
@@ -17,127 +18,140 @@ using namespace shogun;
...
 	bool vfree=false;
 	float64_t* vec=NULL;
 
-	while (changed && (iter<max_iter))
+	for(iter=0; iter<max_iter; iter++)
 	{
-		iter++;
 		if (iter==max_iter-1)
-			SG_SWARNING("kmeans clustering changed throughout %d iterations stopping...\n", max_iter-1)
+			SG_SWARNING("kmeans clustering has reached maximum number of ( %d ) iterations \n", iter)
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: Capital K

Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: ....without having converged. Terminating.

File: src/shogun/clustering/KMeansLloydImpl.cpp (hunk 12 above)
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: ok, thanks for the clarification

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 24]
```diff
@@ -17,127 +18,140 @@ using namespace shogun;
...
 
-		if (iter%1000 == 0)
-			SG_SINFO("Iteration[%d/%d]: Assignment of %i patterns changed.\n", iter, max_iter, changed)
 		changed=0;
+		rhs_mus->copy_feature_matrix(mus);
 
+		distance->precompute_rhs_squared_norms();
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: precompute_rhs

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 25]
```diff
@@ -17,127 +18,140 @@ using namespace shogun;
...
+		
+		/* Assigment step : Assign each point to nearest cluster */
+		for (int32_t i=0; i<lhs_size; i++)
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: openMP please :)

Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: (but in a next patch)

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 26]
```diff
@@ -17,127 +18,140 @@ using namespace shogun;
...
+			const int32_t cluster_assignments_i=cluster_assignments[i];
+			int32_t min_cl, j;
+			float64_t min_dist, dist;
+			
+			min_cl=0;
+		   	min_dist=distance->distance(i,0);
+			for (j=1; j<k; j++)
+			{
+				dist=distance->distance(i,j);
+				if (dist<min_dist)
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: No I mean min rather than this if block
min_dist=CMath::min(min_dist, dist)

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 27]
```diff
@@ -17,127 +18,140 @@ using namespace shogun;
...
+							mus(j, cluster_assignments_i)=0;
+					}
+					
+				}
+
+				cluster_assignments[i] = min_cl;
+			}
+		}
+		if(changed==0)
+			break;
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: we use 4space tabs

File: src/shogun/clustering/KMeansLloydImpl.cpp [hunk 28]
```diff
@@ -17,127 +18,140 @@ using namespace shogun;
...
-				}
-
-				/* ClList(i)= imini ; */
-				ClList[Pat] = imini;
-			}
-		}
+		if (iter%(max_iter/10) == 0)
+			SG_SINFO("Iteration[%d/%d]: Assignment of %i patterns changed.\n", iter, max_iter, changed)
 	}
+	distance->reset_squared_norms();
```
Comment (karlnapf) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: reset_precompute might be a better name

File: src/shogun/distance/Distance.h [hunk 29]
```diff
@@ -120,6 +120,29 @@ class CDistance : public CSGObject
 			return distance(idx_a, idx_b);
 		}
 
+		/** NOT IMPLEMENTED
+		 * Precompute squared norms from features of right hand side
+		 * WARNING : Make sure to reset squared norms using reset_squared_norms()
+		 * when features feature matrix are changed.
+		 * Implemented ONLY for Euclidean distance
```
Comment (karlnapf) [inline src/shogun/distance/Distance.h]: Dont say this that it is only implemented in euclidean distance

File: src/shogun/distance/Distance.h [hunk 30]
```diff
@@ -120,6 +120,29 @@ class CDistance : public CSGObject
 			return distance(idx_a, idx_b);
 		}
 
+		/** NOT IMPLEMENTED
```
Comment (karlnapf) [inline src/shogun/distance/Distance.h]: Dont say not implemented.
Say this is an empty method that might be overloaded in a sub-class

File: src/shogun/distance/Distance.h [hunk 31]
```diff
@@ -120,6 +120,29 @@ class CDistance : public CSGObject
...
+		 */
+		virtual void precompute_rhs_squared_norms(){}
+
+		/** NOT IMPLEMENTED
+		 * Precompute squared norms from features of left hand side
+		 * WARNING : Make sure to reset squared norms using reset_squared_norms()
+		 * when features or feature matrix are changed.
+		 * Implemented ONLY for Euclidean distance
+		 */
+		virtual void precompute_lhs_squared_norms(){}
```
Comment (karlnapf) [inline src/shogun/distance/Distance.h]: same here. And the name change

File: src/shogun/distance/Distance.h [hunk 32]
```diff
@@ -120,6 +120,29 @@ class CDistance : public CSGObject
...
+	
+		/** NOT IMPLEMENTED
+		 * Reset squared norms for features of both sides
+		 * squared norms should be reset whenever features or feature matrix are changed.
+		 * Implemented ONLY for Euclidean distance
+		 */
```
Comment (karlnapf) [inline src/shogun/distance/Distance.h]: Can you send a seperate patch for adding this method to the base class?

Comment (karlnapf) [inline src/shogun/distance/Distance.h]: Then I can merge it first and this patch here gets smaller. It is already quite large

File: tests/unit/clustering/kmeans_unittest.cc [hunk 33]
```diff
@@ -11,6 +11,7 @@
 #include <shogun/features/DenseFeatures.h>
 #include <shogun/clustering/KMeans.h>
 #include <shogun/distance/EuclideanDistance.h>
+#include <shogun/distributions/classical/GaussianDistribution.h>
```
Comment (karlnapf) [inline tests/unit/clustering/kmeans_unittest.cc]: can be removed

File: tests/unit/clustering/kmeans_unittest.cc [hunk 34]
```diff
@@ -162,3 +165,111 @@ TEST(KMeans, minibatch_training_test)
...
+	initial_centers(1,1)=4;
+
+	CDenseFeatures<float64_t>* features=new CDenseFeatures<float64_t>(rect);
+	SG_REF(features);	
+	CEuclideanDistance* distance=new CEuclideanDistance(features, features);
+	CKMeans* clustering=new CKMeans(2, distance,initial_centers);
+	clustering->set_fixed_centers(true);
+
+	clustering->train(features);
+	CDenseFeatures<float64_t>* learnt_centers=CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
```
Comment (karlnapf) [inline tests/unit/clustering/kmeans_unittest.cc]: minor: in c++ you can cast.
This might return a NULL pointer if you do it this way. I guess our static code analysis might complain here.
So better put in an ASSERT_NE

File: tests/unit/clustering/kmeans_unittest.cc [hunk 35]
```diff
@@ -162,3 +165,111 @@ TEST(KMeans, minibatch_training_test)
...
+	SGMatrix<float64_t> c=learnt_centers->get_feature_matrix();
+		
+	EXPECT_EQ(c(0,0), 0);
+	EXPECT_EQ(c(1,0), 5);
+	EXPECT_EQ(c(0,1), 20);
```
Comment (karlnapf) [inline tests/unit/clustering/kmeans_unittest.cc]: this unit test is much better :)

File: tests/unit/clustering/kmeans_unittest.cc [hunk 36]
```diff
@@ -162,3 +165,111 @@ TEST(KMeans, minibatch_training_test)
...
+	rect(0,2)=20;
+	rect(1,2)=20;
+
+	CDenseFeatures<float64_t>* features=new CDenseFeatures<float64_t>(rect);
+	SG_REF(features);	
+	CEuclideanDistance* distance=new CEuclideanDistance(features, features);
+	CKMeans* clustering=new CKMeans(3, distance);
+
+	clustering->train(features);
+	CDenseFeatures<float64_t>* learnt_centers=CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
```
Comment (karlnapf) [inline tests/unit/clustering/kmeans_unittest.cc]: cast and null check

**Comment by karlnapf** [comment]
Great, this patch is evolving nicely. All comments were minor.

Could you send a seperate patch for the CDistance change?
Then we can merge this one. The next one should be openmp for the loop over all data :)

**Comment by karlnapf** [comment]
Travis is also green, so this can be merged soon

File: src/shogun/clustering/KMeansLloydImpl.cpp (hunk 26 above)
Comment (Saurabh7) [inline src/shogun/clustering/KMeansLloydImpl.cpp]: just realized while doing this , need the if block for the cluster assignment anyways ...

**Comment by Saurabh7** [comment]
Just updated with i believe all comments, 
if looks good i could squash and force push again :)

**Comment by karlnapf** [comment]
Looks good to me. Squash!

**Comment by karlnapf** [comment]
BTW this is now the fast version?
Next step: add this to benchmark!

**Comment by karlnapf** [comment]
I leave merging to someone else so that we have another pair of eyes on it
@vigsterkr @lisitsyn @lambday

**Comment by karlnapf** [comment]
Great. Make sure to add your changes to the scripts in these benchmarks. You might want to change the tolerance parameter of scikit. 

Also, we really want to collect these results somewhere to document outcomes of the improvements

**Comment by karlnapf** [comment]
BTW another next step would be to parallelise this thing using openMP. Use google to find inspiration of how to best do this.

**Comment by Saurabh7** [comment]
Sure ! am looking into it
btw heres the output of the benchmark framework with old and new kmeans:
`make run BLOCK=scikit,shogun METHODBLOCK=KMEANS`
old: https://gist.github.com/Saurabh7/a1d271fa52767a7a5344
new: https://gist.github.com/Saurabh7/863c4110b587f31e8eff

for datasets, runtimes > 1 sec:
 `        mlpack  matlab    scikit  mlpy     shogun  weka `
`isolet        -       -  5.112599     -  13.140933     - `
`mnist                  -       -  71.201869     -  141.655496     -  `
`corel-histogram        -       -   1.045997     -    4.541578     - `

`         mlpack  matlab    scikit  mlpy    shogun  weka `
`isolet        -       -  5.102792     -  4.763088     - `
`mnist                  -       -  71.451103     -  49.875730     - `
`corel-histogram        -       -   1.052271     -   2.439281     -`

**Comment by karlnapf** [comment]
The second is new? Which means we are now quite a bit faster. It would be good to check why we are faster in certain cases, and not in others. But anyways -- make sure that all these tests are going back to the benchmark system. Reference this PR when you send them a patch

**Comment by karlnapf** [comment]
Ok I will merge this. Great work!

BTW do we have a notebook for KMeans?
BTW2 maybe next check how to use openmp in kmeans ? I found a few clever implementations and it might be worth to include this here too.

**Comment by Saurabh7** [comment]
yup we have a notebook made back then http://shogun-toolbox.org/static/notebook/current/KMeans.html.
yes I am having a look into openmp stuff now !
